### PR TITLE
refactor(stream): use columnar eval for temporal join non-lookup conds

### PR DESCRIPTION
**Pull request by TennyZhuang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Will do a benchmark later

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by TennyZhuang** [comment]
Note: Check "Hide whitespace" can significantly improve your review experience.
image

**Review by wangrunji0408 — COMMENTED** [review]
Generally LGTM

File: src/stream/src/executor/temporal_join.rs [hunk 1]
```diff
@@ -400,57 +401,96 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> TemporalJoinExecutor
...
-                                } else {
-                                    true
+                    // Joined result without evaluating other conditions.
+                    let st1 = {
+                        #[try_stream]
+                        async {
+                            #[allow(unreachable_code)]
+                            if false {
+                                return Err(unreachable!("type hints only") as StreamExecutorError);
+                            }
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/stream/src/executor/temporal_join.rs:411]: It's non-trivial to declare a variable as `impl Trait` type. TAIT must be enabled for that, and there are many corner cases. So I guess the macro author don't want to do that.

File: src/stream/src/executor/temporal_join.rs [hunk 2]
```diff
@@ -400,57 +401,96 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> TemporalJoinExecutor
...
+                            assert!(chunk.visibility().all());
+                            // For non matched row, we shouldn't evaluate on it.
+                            // So we treat `row_matched` as visibility here.
+                            let chunk = chunk.clone_with_vis(row_matched.clone());
+                            let (data_chunk, ops) = chunk.into_parts();
+                            let filter = cond.eval_infallible(&data_chunk).await;
+                            let ArrayImpl::Bool(bool_array) = &*filter else {
+                                panic!("unmatched type: filter expr returns a non-null array");
+                            };
+                            let new_vis = bool_array.to_bitmap() | (!row_matched);
```
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/stream/src/executor/temporal_join.rs:487]: LEFT OUTER JOIN

[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/stream/src/executor/temporal_join.rs:487]: Non-match rows will only be appended to the chunk when it's a LEFT OUTER join, so they should visible here.

File: src/stream/src/executor/temporal_join.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/stream/src/executor/temporal_join.rs:411]: I'll create a helper macro to generate the code.

**Comment by TennyZhuang** [comment]
Need to handle `i2o_mapping` after the second phase.

**Comment by yuhao-su** [comment]
@chenzl25 Is it possible the LHS be non append only in the future? If yes doing columnar eval can be more complicated then

**Comment by yuhao-su** [comment]
If LHS matches a row in RHS and in `st1`, but the row is filter out in `st2`, we should expect a matched empty row in the result.

File: src/stream/src/executor/temporal_join.rs [hunk 3]
```diff
@@ -400,57 +412,117 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> TemporalJoinExecutor
...
-                                    true
+                    // Joined result without evaluating non-lookup conditions.
+                    let st1 = {
+                        #[try_stream]
+                        async {
+                            #[allow(unreachable_code)]
+                            #[allow(clippy::diverging_sub_expression)]
+                            if false {
+                                return unreachable!("type hints only") as StreamExecutorResult<_>;
+                            }
```
Comment (stdrc) [inline src/stream/src/executor/temporal_join.rs:423]: Is it possible to extract a method for this async block? The hack here looks ugly🤣

**Comment by TennyZhuang** [comment]
> If LHS matches a row in RHS and in `st1`, but the row is filter out in `st2`, we should expect a matched empty row in the result.

Recorded in #15257 

The PR doesn't make the behavior worse, I guess we can merge it first.

File: src/stream/src/executor/temporal_join.rs (hunk 3 above)
Comment (TennyZhuang) [reply src/stream/src/executor/temporal_join.rs:423]: Prefer to keep the hack code in a small scope and use it only needed.

**Comment by chenzl25** [comment]
> @chenzl25 Is it possible the LHS be non append only in the future? If yes doing columnar eval can be more complicated then

Yes, we do want to support non-append only temporal join in the future. https://github.com/risingwavelabs/risingwave/issues/15218

**Review by yuhao-su — APPROVED** [review]

File: src/stream/src/executor/temporal_join.rs (hunk 3 above)
Comment (xxchan) [reply src/stream/src/executor/temporal_join.rs:423]: I think this is very acceptable. I like the simple and stupid way

**Review by xxchan — APPROVED** [review]

**Comment by TennyZhuang** [comment]
image
Nexmark result with an simple non-lookup condition.

**Pull request merged**